src/rpc/xmlrpc_tinyxml2.cc

```diff
@@ -0,0 +1,451 @@
+  if (delim_pos == target_string.npos ||
+      delim_pos + 2 >= target_string.size()) {
+	if (require_index) {
+      throw torrent::input_error("invalid parameters: no index");
+    }
+    hash = target_string;
+  } else {
+    hash  = target_string.substr(0, delim_pos);
+    type  = target_string[delim_pos + 1];
+    index = target_string.substr(delim_pos + 2);
+  }
+  core::Download* download = xmlrpc.slot_find_download()(hash.c_str());
+
+  if (download == nullptr)
+    throw torrent::input_error("invalid parameters: info-hash not found");
+
+  try {
+    switch (type) {
+      case 'd':
+        *target = rpc::make_target(download);
+        break;
+      case 'f':
+        *target = rpc::make_target(
+          command_base::target_file,
+          xmlrpc.slot_find_file()(download, std::stoi(std::string(index))));
+        break;
+      case 't':
+        *target = rpc::make_target(
+          command_base::target_tracker,
+          xmlrpc.slot_find_tracker()(download, std::stoi(std::string(index))));
+        break;
+      case 'p': {
+          if (index.size() < 40) {
+            throw xmlrpc_error(XMLRPC_TYPE_ERROR, "Not a hash string.");
+          }
+          torrent::HashString hash;
+          torrent::hash_string_from_hex_c_str(index.c_str(), hash);
+          *target = rpc::make_target(
+                                     command_base::target_peer,
+                                     xmlrpc.slot_find_peer()(download, hash));
+          break;
+      }
+      default:
+        throw torrent::input_error("invalid parameters: unexpected target type");
+    }
+  } catch (const std::logic_error&) {
+    throw torrent::input_error("invalid parameters: invalid index");
+  }
+}
+
+torrent::Object execute_command(std::string method_name, const tinyxml2::XMLElement* params_element) {
+  if (params_element == nullptr)
+    throw xmlrpc_error(XMLRPC_INTERNAL_ERROR, "invalid parameters: null");
+  CommandMap::iterator cmd_itr = commands.find(method_name.c_str());
+  if (cmd_itr == commands.end() || !(cmd_itr->second.m_flags & CommandMap::flag_public_xmlrpc)) {
+    throw xmlrpc_error(XMLRPC_NO_SUCH_METHOD_ERROR, "Method '" + std::string(method_name) + "' not defined");
+  }
+  torrent::Object params_raw = torrent::Object::create_list();
+  torrent::Object::list_type& params = params_raw.as_list();
+  if (params_element != nullptr) {
+    for (auto child = params_element->FirstChildElement("param"); child; child = child->NextSiblingElement("param")) {
+      params.push_back(xml_value_to_object(child->FirstChildElement("value")));
+    }
+  }
+  rpc::target_type target = rpc::make_target();
+  if (params.size() == 0 && (cmd_itr->second.m_flags & (CommandMap::flag_file_target | CommandMap::flag_tracker_target))) {
+    throw xmlrpc_error(XMLRPC_TYPE_ERROR, "invalid parameters: too few");
+  }
+  if (params.size() > 0) {
+    object_to_target(params.front(), cmd_itr->second.m_flags, &target);
+    params.erase(params.begin());
```

Comment:
Instead of erasing the target param here, just store it separately from the start.